File: services/23-symfony/src/Controller/AuthController.php
<?php

namespace App\Controller;

use App\Entity\User;
use App\Repository\UserRepository;
use Doctrine\ORM\EntityManagerInterface;
use Lexik\Bundle\JWTAuthenticationBundle\Services\JWTTokenManagerInterface;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;
use Symfony\Component\Security\Http\Attribute\CurrentUser;

#[Route('/auth')]
class AuthController extends AbstractController
{
    public function __construct(
        private readonly UserRepository $userRepository,
        private readonly EntityManagerInterface $em,
        private readonly JWTTokenManagerInterface $jwtManager,
    ) {}

    /** GET /auth/login — redirect to GitHub OAuth. */
    #[Route('/login', methods: ['GET'])]
    public function login(Request $request): Response
    {
        $clientId    = $_ENV['GITHUB_CLIENT_ID']    ?? '';
        $redirectUri = $_ENV['GITHUB_REDIRECT_URI'] ?? '';
        $url = 'https://github.com/login/oauth/authorize'
             . '?client_id=' . urlencode($clientId)
             . '&redirect_uri=' . urlencode($redirectUri)
             . '&scope=user:email';
        return $this->redirect($url);
    }

    /** GET /auth/callback — exchange code, upsert user, return JWT. */
    #[Route('/callback', methods: ['GET'])]
    public function callback(Request $request): JsonResponse
    {
        $code = $request->query->get('code');
        if (empty($code)) {
            return $this->json(['error' => 'Missing code'], 400);
        }

        // Exchange code for access token.
        $ch = curl_init('https://github.com/login/oauth/access_token');
        curl_setopt_array($ch, [
            CURLOPT_POST           => true,
            CURLOPT_POSTFIELDS     => http_build_query([
                'client_id'     => $_ENV['GITHUB_CLIENT_ID']     ?? '',
                'client_secret' => $_ENV['GITHUB_CLIENT_SECRET'] ?? '',
                'code'          => $code,
            ]),
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER     => ['Accept: application/json'],
        ]);
        $tokenBody = json_decode(curl_exec($ch), true) ?? [];
        curl_close($ch);

        $accessToken = $tokenBody['access_token'] ?? null;
        if (empty($accessToken)) {
            return $this->json(['error' => 'GitHub token exchange failed'], 401);
        }

        // Fetch GitHub user.
        $ch = curl_init('https://api.github.com/user');
        curl_setopt_array($ch, [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER     => [
                "Authorization: Bearer $accessToken",
                'Accept: application/json',
                'User-Agent: symfony-app',
            ],
        ]);
        $profile = json_decode(curl_exec($ch), true) ?? null;
        curl_close($ch);

        if ($profile === null) {
            return $this->json(['error' => 'GitHub profile fetch failed'], 401);
        }

        $email = $profile['email'] ?? ($profile['login'] . '@github.noemail');
        $name  = $profile['name']  ?? $profile['login'];

        // Upsert user.
        $user = $this->userRepository->findOneBy(['email' => $email]);
        if ($user === null) {
            $user = new User($email, $name, 'github');
            $this->em->persist($user);
        } else {
            $user->setName($name);
        }
        $this->em->flush();

        $token = $this->jwtManager->createFromPayload($user, [
            'sub'   => $user->getId(),
            'email' => $user->getEmail(),
            'name'  => $user->getName(),
        ]);
        return $this->json(['token' => $token]);
    }

    /** GET /auth/me — return current user (authenticated). */
    #[Route('/me', methods: ['GET'])]
    public function me(#[CurrentUser] ?User $user): JsonResponse
    {
        if ($user === null) return $this->json(['error' => 'Unauthorized'], 401);
        return $this->json([
            'id'    => $user->getId(),
            'email' => $user->getEmail(),
            'name'  => $user->getName(),
        ]);
    }

    /** POST /auth/logout — stateless JWT. */
    #[Route('/logout', methods: ['POST'])]
    public function logout(): JsonResponse
    {
        return $this->json(['status' => 'ok']);
    }

    /** POST /auth/api-key — generate and save API key. */
    #[Route('/api-key', methods: ['POST'])]
    public function generateApiKey(#[CurrentUser] ?User $user): JsonResponse
    {
        if ($user === null) return $this->json(['error' => 'Unauthorized'], 401);
        $key = bin2hex(random_bytes(32));
        $user->setApiKey($key);
        $this->em->flush();
        return $this->json(['api_key' => $key]);
    }

    /** DELETE /auth/api-key — revoke API key. */
    #[Route('/api-key', methods: ['DELETE'])]
    public function revokeApiKey(#[CurrentUser] ?User $user): JsonResponse
    {
        if ($user === null) return $this->json(['error' => 'Unauthorized'], 401);
        $user->setApiKey(null);
        $this->em->flush();
        return $this->json(['status' => 'ok']);
    }

    /** POST /dev/token — dev-only JWT without OAuth. */
    #[Route('/dev/token', methods: ['POST'])]
    public function devToken(Request $request): JsonResponse
    {
        $env = $_ENV['APP_ENV'] ?? 'dev';
        if ($env === 'prod') {
            return $this->json(['error' => 'Not available'], 403);
        }

        $body  = json_decode($request->getContent(), true) ?? [];
        $email = $body['email'] ?? 'dev@example.com';
        $name  = $body['name']  ?? 'Dev User';

        $user = $this->userRepository->findOneBy(['email' => $email]);
        if ($user === null) {
            $user = new User($email, $name, 'dev');
            $this->em->persist($user);
            $this->em->flush();
        }

        $token = $this->jwtManager->createFromPayload($user, [
            'sub'   => $user->getId(),
            'email' => $user->getEmail(),
            'name'  => $user->getName(),
        ]);
        return $this->json(['token' => $token]);
    }
}

File: services/17-micronaut/src/main/java/com/example/controller/AuthController.java
package com.example.controller;

import com.example.auth.JwtAuthenticator;
import com.example.entity.User;
import com.example.repository.UserRepository;
import io.micronaut.context.annotation.Value;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.annotation.*;
import io.micronaut.security.annotation.Secured;
import io.micronaut.security.authentication.Authentication;
import io.micronaut.security.rules.SecurityRule;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

@Controller
public class AuthController {

    private final UserRepository userRepository;
    private final JwtAuthenticator jwtAuthenticator;

    @Value("${github.client.id:}")
    private String githubClientId;

    @Value("${github.client.secret:}")
    private String githubClientSecret;

    @Value("${github.redirect.uri:http://localhost:8080/auth/callback}")
    private String githubRedirectUri;

    public AuthController(UserRepository userRepository, JwtAuthenticator jwtAuthenticator) {
        this.userRepository = userRepository;
        this.jwtAuthenticator = jwtAuthenticator;
    }

    /** GET /auth/login — redirect to GitHub OAuth. */
    @Get("/auth/login")
    @Secured(SecurityRule.IS_ANONYMOUS)
    public HttpResponse<?> login() {
        String url = "https://github.com/login/oauth/authorize"
                + "?client_id=" + URLEncoder.encode(githubClientId, StandardCharsets.UTF_8)
                + "&redirect_uri=" + URLEncoder.encode(githubRedirectUri, StandardCharsets.UTF_8)
                + "&scope=user:email";
        return HttpResponse.redirect(URI.create(url));
    }

    /** GET /auth/callback — exchange code, upsert user, return JWT. */
    @Get("/auth/callback")
    @Secured(SecurityRule.IS_ANONYMOUS)
    public HttpResponse<Map<String, String>> callback(@QueryValue String code) {
        // Exchange code for GitHub access token.
        // Simplified: use java.net.http — Micronaut HTTP client also available.
        try {
            var form = "client_id=" + githubClientId
                    + "&client_secret=" + githubClientSecret
                    + "&code=" + code;
            var tokenReq = HttpRequest.newBuilder()
                    .uri(URI.create("https://github.com/login/oauth/access_token"))
                    .header("Accept", "application/json")
                    .header("Content-Type", "application/x-www-form-urlencoded")
                    .POST(HttpRequest.BodyPublishers.ofString(form))
                    .build();
            var tokenRes = HttpClient.newHttpClient().send(tokenReq,
                    java.net.http.HttpResponse.BodyHandlers.ofString());
            // Parse access_token from JSON response.
            String body = tokenRes.body();
            String accessToken = extractJson(body, "access_token");
            if (accessToken == null) {
                return HttpResponse.unauthorized();
            }

            // Fetch GitHub user.
            var profileReq = HttpRequest.newBuilder()
                    .uri(URI.create("https://api.github.com/user"))
                    .header("Authorization", "Bearer " + accessToken)
                    .header("Accept", "application/json")
                    .GET().build();
            var profileRes = HttpClient.newHttpClient().send(profileReq,
                    java.net.http.HttpResponse.BodyHandlers.ofString());
            String profile = profileRes.body();

            String email = extractJson(profile, "email");
            String login = extractJson(profile, "login");
            String name = extractJson(profile, "name");
            if (email == null || email.equals("null")) {
                email = login + "@github.noemail";
            }
            if (name == null || name.equals("null")) name = login;

            final String finalEmail = email;
            final String finalName = name;

            User user = userRepository.findByEmail(email).orElseGet(() -> {
                User u = new User(finalEmail, finalName, "github");
                return userRepository.save(u);
            });
            if (!user.getName().equals(finalName)) {
                user.setName(finalName);
                user = userRepository.update(user);
            }

            String jwt = jwtAuthenticator.generateToken(user.getId(), user.getEmail(), user.getName());
            return HttpResponse.ok(Map.of("token", jwt));
        } catch (Exception e) {
            return HttpResponse.serverError(Map.of("error", "OAuth flow failed"));
        }
    }

    /** GET /auth/me — return current user profile. */
    @Get("/auth/me")
    @Secured(SecurityRule.IS_AUTHENTICATED)
    public HttpResponse<Map<String, Object>> me(Authentication authentication) {
        String userId = (String) authentication.getAttributes().get("sub");
        String email = (String) authentication.getAttributes().get("email");
        String name = (String) authentication.getAttributes().get("name");
        return HttpResponse.ok(Map.of("id", userId, "email", email, "name", name));
    }

    /** POST /auth/logout — stateless JWT; no server-side action needed. */
    @Post("/auth/logout")
    @Secured(SecurityRule.IS_AUTHENTICATED)
    public HttpResponse<Map<String, String>> logout() {
        return HttpResponse.ok(Map.of("status", "ok"));
    }

    /** POST /auth/api-key — generate and save an API key. */
    @Post("/auth/api-key")
    @Secured(SecurityRule.IS_AUTHENTICATED)
    public HttpResponse<Map<String, String>> generateApiKey(Authentication authentication) {
        UUID userId = UUID.fromString((String) authentication.getAttributes().get("sub"));
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        String apiKey = "sk_" + UUID.randomUUID().toString().replace("-", "");
        user.setApiKey(apiKey);
        userRepository.update(user);
        return HttpResponse.ok(Map.of("apiKey", apiKey));
    }

    /** DELETE /auth/api-key — revoke the current user's API key. */
    @Delete("/auth/api-key")
    @Secured(SecurityRule.IS_AUTHENTICATED)
    public HttpResponse<Void> deleteApiKey(Authentication authentication) {
        UUID userId = UUID.fromString((String) authentication.getAttributes().get("sub"));
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        user.setApiKey(null);
        userRepository.update(user);
        return HttpResponse.ok();
    }

    /** POST /dev/token — dev-only; generates JWT without OAuth. */
    @Post("/dev/token")
    @Secured(SecurityRule.IS_ANONYMOUS)
    public HttpResponse<Map<String, String>> devToken(
            @QueryValue(defaultValue = "dev@example.com") String email,
            @QueryValue(defaultValue = "Dev User") String name) {

        String env = System.getenv("MICRONAUT_ENV");
        if ("production".equals(env)) {
            return HttpResponse.notFound();
        }

        User user = userRepository.findByEmail(email).orElseGet(() -> {
            User u = new User(email, name, "dev");
            return userRepository.save(u);
        });

        String jwt = jwtAuthenticator.generateToken(user.getId(), user.getEmail(), user.getName());
        return HttpResponse.ok(Map.of("token", jwt));
    }

    // ── Helpers ────────────────────────────────────────────────────────────

    /** Minimal JSON field extractor — avoids pulling in a full JSON library. */
    private String extractJson(String json, String key) {
        String pattern = "\"" + key + "\":";
        int idx = json.indexOf(pattern);
        if (idx < 0) return null;
        int start = idx + pattern.length();
        // Skip whitespace.
        while (start < json.length() && json.charAt(start) == ' ') start++;
        if (start >= json.length()) return null;
        if (json.charAt(start) == '"') {
            int end = json.indexOf('"', start + 1);
            return end < 0 ? null : json.substring(start + 1, end);
        }
        if (json.startsWith("null", start)) return null;
        int end = start;
        while (end < json.length() && json.charAt(end) != ',' && json.charAt(end) != '}') end++;
        return json.substring(start, end).trim();
    }
}
